File: C++/Graphs/Multisource_Shortest_Path.cpp
#include<bits/stdc++.h>
#define fast_io ios_base::sync_with_stdio(false);cin.tie(NULL)
using namespace std;

void solver() {
    cout << "Enter the number of vertices,edges and sources" << endl;
    
    int n,m,s;
    cin >> n >> m >> s;
  
    vector<int> adj[n+1];
    vector<int> sources;
    
    cout << "Enter the sources: ";
    
    for(int i = 0;i<s;i++) {
        int src;
        cin >> src;
        
        sources.push_back(src);
    }
    
    cout << "Enter the edges" << endl;
    
    while(m--) {
        int u,v;
        cin >> u >> v;
        
        adj[u].push_back(v);
        adj[v].push_back(u);
    }
    
    vector<int> dist(n+1,1e9);
    vector<int> vis(n+1,0);
    queue<int> q;
    
    for(auto it: sources) {
        vis[it] = 1;
        q.push(it);
    }
    
    int lvl = 0;
    
    while(q.size()) {
        int sz = q.size();
        while(sz--) {
            int node = q.front();
            q.pop();
            
            dist[node] = lvl;
            
            for(auto it: adj[node]) {
                if(vis[it]) {
                    continue;
                }
                
                vis[it] = 1;
                q.push(it);
            }
        }
        lvl++;
        
    }
    
    for(int i = 1;i<=n;i++) {
        cout << i << ": " << dist[i] << endl;
    }
}

int main() {
    fast_io;
    int t = 1;
    
    while(t--) {
       solver();
    }
    
    return 0;
}

File: Python/Graphs/MultiSource_Shortest_Path.py
from collections import deque

print("Enter the number of vertices, edges, and sources")
n, m, s = map(int, input().split())
adj = [[] for _ in range(n+1)]
sources = input("Enter the sources ").split()

print("Enter the edges:")

while m > 0:
    u, v = map(int, input().split())
    
    adj[u].append(v)
    adj[v].append(u)
    
    m -= 1

dist = [1e9] * (n + 1)
vis = [0] * (n + 1)
q = deque()

for it in sources:
    it = int(it)
    vis[it] = 1
    q.append(it)

lvl = 0

while q:
    sz = len(q)
    while sz > 0:
        node = q.popleft()
        
        dist[node] = lvl
        
        for it in adj[node]:
            if vis[it]:
                continue
            
            vis[it] = 1
            q.append(it)
        
        sz -= 1
    
    lvl += 1

for i in range(1, n+1):
    print(i, ":", dist[i])